src/vma/lwip/tcp_in.c

```diff
@@ -341,19 +341,36 @@ tcp_listen_input(struct tcp_pcb_listen *pcb, tcp_in_data* in_data)
     return ERR_OK;
   }
 
-  if (in_data->flags & TCP_FIN) {
-    /* An incoming FIN should be ignored. Return. */
-    return ERR_OK;
-  }
-
   /* In the LISTEN state, we check for incoming SYN segments,
      creates a new PCB, and responds with a SYN|ACK. */
   if (in_data->flags & TCP_ACK) {
     /* For incoming segments with the ACK flag set, respond with a
        RST. */
     LWIP_DEBUGF(TCP_RST_DEBUG, ("tcp_listen_input: ACK in LISTEN, sending reset\n"));
-    tcp_rst(in_data->ackno + 1, in_data->seqno + in_data->tcplen,
-      in_data->tcphdr->dest, in_data->tcphdr->src, NULL);
+    rc = tcp_rst(in_data->ackno + 1, in_data->seqno + in_data->tcplen,
+      in_data->tcphdr->dest, in_data->tcphdr->src, pcb);
+    if (rc < 0) {
+      struct tcp_hdr tcphdr, *ptcp_hdr = (struct tcp_hdr*)&tcphdr;
+      TCP_EVENT_CLONE_PCB(pcb, &npcb, ERR_OK, rc);
+      if (npcb == NULL) {
+        LWIP_DEBUGF(TCP_DEBUG, ("tcp_listen_input: could not allocate PCB\n"));
+        TCP_STATS_INC(tcp.memerr);
+        return ERR_MEM;
+      }
+      ip_addr_copy(npcb->local_ip, in_data->iphdr->dest);
+      ip_addr_copy(npcb->remote_ip, in_data->iphdr->src);
+      npcb->remote_port = in_data->tcphdr->src;
+      ptcp_hdr->src = htons(in_data->tcphdr->dest);
+      ptcp_hdr->dest = htons(in_data->tcphdr->src);
+      ptcp_hdr->seqno = htonl(in_data->ackno);
+      ptcp_hdr->ackno = 0;
+      TCPH_HDRLEN_FLAGS_SET(ptcp_hdr, TCP_HLEN/4, TCP_RST);
+      ptcp_hdr->wnd = 0;
+      ptcp_hdr->chksum = 0;
+      ptcp_hdr->urgp = 0;
+      npcb->callback_arg = (void*)ptcp_hdr;
+      TCP_DST_NC_SEND(pcb, npcb, ERR_OK, rc);
+    }
   } else if (in_data->flags & TCP_SYN) {
```

Comment:
Most of the logic is duplicated from tcp_rst.
why can't we create a temporary pcb and call tcp_rst with it to send the data ?

Reply:
I agree, it can be optimized. Thanks.